core/src/OpenACC/Kokkos_OpenACC_Team.hpp

```diff
@@ -0,0 +1,616 @@
+
+#include <openacc.h>
+#include <impl/Kokkos_Traits.hpp>
+#include <impl/Kokkos_Spinwait.hpp>
+#include <Kokkos_Atomic.hpp>
+#include <OpenACC/Kokkos_OpenACC.hpp>
+#include <Kokkos_ExecPolicy.hpp>
+
+//----------------------------------------------------------------------------
+//----------------------------------------------------------------------------
+
+namespace Kokkos {
+namespace Impl {
+
+class OpenACCTeamMember {
+ public:
+  constexpr static int TEAM_REDUCE_SIZE      = 512;
+  constexpr static int DEFAULT_TEAM_SIZE_MAX = 512;
+  constexpr static int DEFAULT_TEAM_SIZE_REC = 128;
+
+  using execution_space      = Kokkos::Experimental::OpenACC;
+  using scratch_memory_space = execution_space::scratch_memory_space;
+
+  scratch_memory_space m_team_shared;
+  int m_team_scratch_size[2];
+  int m_team_rank;
+  int m_team_size;
+  int m_league_rank;
+  int m_league_size;
+  int m_vector_length;
+  int m_vector_lane;
+  int m_shmem_block_index;
+  void* m_glb_scratch;
+  void* m_reduce_scratch;
+
+ public:
+  KOKKOS_INLINE_FUNCTION
+  const execution_space::scratch_memory_space& team_shmem() const {
+    return m_team_shared.set_team_thread_mode(0, 1, 0);
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  const execution_space::scratch_memory_space& team_scratch(int level) const {
+    return m_team_shared.set_team_thread_mode(level, 1,
+                                              m_team_scratch_size[level]);
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  const execution_space::scratch_memory_space& thread_scratch(int level) const {
+    return m_team_shared.set_team_thread_mode(level, team_size(), team_rank());
+  }
+
+  KOKKOS_INLINE_FUNCTION int league_rank() const { return m_league_rank; }
+  KOKKOS_INLINE_FUNCTION int league_size() const { return m_league_size; }
+  KOKKOS_INLINE_FUNCTION int team_rank() const {
+#ifdef KOKKOS_COMPILER_NVHPC
+    return __pgi_vectoridx();
+#else
+    return m_team_rank;
+#endif
+  }
+  KOKKOS_INLINE_FUNCTION int vector_length() const { return m_vector_length; }
+  KOKKOS_INLINE_FUNCTION int team_size() const { return m_team_size; }
+  KOKKOS_INLINE_FUNCTION void* impl_reduce_scratch() const {
+    return m_reduce_scratch;
+  }
+
+  // FIXME_OPENACC: OpenACC does not provide any explicit barrier constructs
+  // for device kernels.
+  KOKKOS_INLINE_FUNCTION void team_barrier() const {
+    Kokkos::abort(
+        std::string(
+            "Kokkos::Experimental::OpenACC ERROR: OpenACC does not "
+            "provide any explicit barrier constructs for device kernels; exit!")
+            .c_str());
+    exit(0);
+  }
+
+  // FIXME_OPENACC: team_broadcast() is not implemented.
+  template <class ValueType>
+  KOKKOS_INLINE_FUNCTION void team_broadcast(ValueType& value,
+                                             int thread_id) const {
+    static_assert(std::is_void_v<ValueType>,
+                  "Kokkos Error: team_broadcast() is not implemented for the "
+                  "OpenACC backend");
```

Comment:
why only fail compilation if ValueType is void? the comment reads as if we can never call this function

Reply:
Isn't it the opposite? It only fails if `ValueType` is **not** void. It would be clearer using `static_assert(false, ...)`

Reply:
The intent is that it always fails at compile time if the function is instantiated.
`static_assert(false)` would be a hard compile error regardless of whether it is instantiated.
I had suggested that trick in the past.  Admittedly this does not convey the intent.  Maybe we can introduce `always_false_v<T>`.  I already had added some `always_true` trait.

Reply:
> Isn't it the opposite? It only fails if `ValueType` is **not** void. It would be clearer using `static_assert(false, ...)`

you are right, sry for the confusion

Reply:
I created `Kokkos::Impl::always_false<T>` in `impl/Kokkos_Utilities.hpp` file, and used it to issue the compile-time error if unimplemented functions are instantiated.